### PR TITLE
[torch.library] Change Library.__del__ into weakref.finalize

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #101829

`__del__` is a bit difficult to use, because when it is called, it is
not guaranteed that anything it uses has not been cleaned up.

Ed tells me he got the following exception one day, which is what
prompted this PR.
```
Exception ignored in: <function Library.__del__ at 0x7fa36d211e50>
Traceback (most recent call last):
  File "/data/users/ezyang/a/pytorch/torch/library.py", line 139, in
  __del__
  AttributeError: 'NoneType' object has no attribute 'remove'
```

One solution is to use weakref.finalize, which lets one define a
function to be run when the object is deleted that can hold references
to specific things it needs.

Another solution is to just check if the object is None, but I like the
weakref solution better.

Test Plan:
- new test

### CODE REVIEW

File: torch/library.py [hunk 1]
```diff
@@ -131,13 +137,10 @@ def impl(self, op_name, fn, dispatch_key=''):
...
-        # _op_impls might not have been initialized if an error was thrown in __init__
-        _op_impls_ = getattr(self, '_op_impls', None)
-        if _op_impls_:
-            for key in self._op_impls:
-                _impls.remove(key)
-            del self.m
+
+def _del_library(captured_impls, op_impls):
+    for key in op_impls:
+        captured_impls.remove(key)
```
Comment (Skylion007) [inline torch/library.py:143]: These are both sets, right?
```suggestion
    captured_impls.difference_update(ops_impl)
```
or simply
```suggestion
    caputred_impls -= ops_impl
```

**Review by ezyang — APPROVED** [review]
This seems OK, but please double check that you are not creating a refcycle through the finalizer. If so, you may want to pass in weakrefs to the finalizer and skip finalization if the weakrefs have gone dead.

File: torch/library.py (hunk 1 above)
Comment (zou3519) [reply torch/library.py:143]: Yes, you're correct

**Comment by zou3519** [comment]
> but please double check that you are not creating a refcycle through the finalizer. 

Verified, and also added new tests to check that the refcounts are what we expect (and that there is no reference cycle as a result)

**Pull request closed without merging**